### PR TITLE
Interval arithmetic operations

### PR DESCRIPTION
**Pull request by Upabjojr**

### CODE REVIEW

**Comment by gxyd** [comment]
As a general comment. There has been some work done in the past which used the operations similar to the *Interval Arithmetic*. For example look at `sympy/calculus/util.py` has `AccumulationBounds` whose operations were inherited from @hargup's PR on interval arithmetic #2682.
Perhaps I have not even looked at your PR changes (little busy with other things).

**Comment by Upabjojr** [comment]
@gxyd oh, I've just noticed. Does `AccumBounds` support open/closed intervals in various combinations? If not, this PR adds something new to SymPy.

**Comment by gxyd** [comment]
The thing is `AccumBounds` are *not* Intervals they can be thought of as 
loose bounds on a function, so as in such they are always closed on the 
ends points (unless `-oo` or `oo` occur) being. My point is arithmetic 
is same as `Interval Arithmetic`. So may be the code could be re-used.

Also I found the paper 
http://fab.cba.mit.edu/classes/S62.12/docs/Hickey_interval.pdf a bit 
more useful than the wikipedia article.

**Comment by Upabjojr** [comment]
> The thing is `AccumBounds` are *not* Intervals they can be thought of as loose bounds on a function, so as in such they are always closed on the ends points (unless `-oo` or `oo` occur) being.

I think we need to add some interval assumptions support, and be able to distinguish conditions like `x > 1` vs `x >= 1`.

> So may be the code could be re-used.

Yes, I notice now that the code is quite similar in certain cases. `AccumulationBounds` are currently more complete (except for the lack of open-interval support).

The point for this PR is that this _arithmetic_ could be generalized to any set (finite sets, naturals, unions, intersections), and maybe eventually extend the assumptions.

**Comment by gxyd** [comment]
Well there is another PR that is currently open 
https://github.com/sympy/sympy/pull/2721 . That PR also states:

 > I was curious if we could do this for all of sets, not just 
Intervals. My solution is |SetExpr|, a class that acts like an |Expr| 
but carries around a |Set| of possible values. This supports general set 
arithmetic of which interval arithmetic is a special case.

which sounds like what you are talking about. Is that right?

**Comment by Upabjojr** [comment]
> Well there is another PR that is currently open #2721 .

Oh, I should have asked in the forum before. That PR looks interesting.

> which sounds like what you are talking about. Is that right?

Yes.

**Comment by gxyd** [comment]
We also had a few discussions on the @jksuom's sympy gitter channel, 
starting with this message 
https://gitter.im/jksuom/sympy?at=562f173b70e3dd4b5039cf4a about what we 
had in mind while implementing "Accumulation Bounds".

Yes, I also believe asking on forum could generate some useful input on 
the topic.

**Comment by asmeurer** [comment]
Is the idea here to eventually add things like `SetAdd` (where `SetAdd(A, B) = A + B = {a + b | a in A, b in B}`)? I think that level of formalism would be useful to have.

**Comment by Upabjojr** [comment]
> Is the idea here to eventually add things like SetAdd (where SetAdd(A, B) = A + B = {a + b | a in A, b in B})? I think that level of formalism would be useful to have. 

Yes, that was the idea. But... as pointed out by @gxyd , it was already object of a PR 3 years ago: https://github.com/sympy/sympy/pull/2721

Why didn't it get merged?

File: sympy/sets/sets.py [hunk 1]
```diff
@@ -980,6 +980,82 @@ def _contains(self, other):
...
+            return Interval(self.start + other.start, self.end + other.end, self.left_open | other.left_open, self.right_open | other.right_open)
+        raise NotImplementedError()
+
+    def _sub(self, other):
+        """
+        Subtractions in interval arithmetic
+        https://en.wikipedia.org/wiki/Interval_arithmetic
+        """
+        if isinstance(other, Interval):
+            return Interval(self.start - other.end, self.end - other.start, self.left_open | other.right_open, self.right_open | other.left_open)
```
Comment (asmeurer) [inline sympy/sets/sets.py]: Are you using `|` instead of `or` here intentionally? Can these be symbolic?

Comment (Upabjojr) [inline sympy/sets/sets.py]: Uhm... I didn't pay much attention.

Comment (asmeurer) [inline sympy/sets/sets.py]: `or` is better, unless you explicitly want to create a symbolic object. It's more readable, and can't accidentally do bitwise or.

Comment (asmeurer) [inline sympy/sets/sets.py]: I'd say this note applies to bitwise operators basically http://docs.sympy.org/latest/modules/logic.html#sympy.logic.boolalg.BooleanTrue

**Pull request closed without merging**